File: ODK.Website.Angular/src/app/services/emails/email-admin.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

import { catchApiError } from '../http/catchApiError';
import { ChapterEmail } from 'src/app/core/emails/chapter-email';
import { ChapterEmailProvider } from 'src/app/core/emails/chapter-email-provider';
import { Email } from 'src/app/core/emails/email';
import { EmailType } from 'src/app/core/emails/email-type';
import { environment } from 'src/environments/environment';
import { HttpUtils } from '../http/http-utils';
import { ServiceResult } from '../service-result';
import { StringUtils } from 'src/app/utils/string-utils';

const baseUrl = `${environment.adminApiBaseUrl}/emails`;

const endpoints = {
  chapterEmail: (chapterId: string, type: EmailType) => `${baseUrl}/chapters/${chapterId}/${EmailType[type]}`,
  chapterEmails: (id: string) => `${baseUrl}/chapters/${id}`,
  chapterEmailProvider: (id: string, chapterEmailProviderId: string) => `${baseUrl}/chapters/${id}/providers/${chapterEmailProviderId}`,
  chapterEmailProviders: (id: string) => `${baseUrl}/chapters/${id}/providers`,
  email: (chapterId: string, type: EmailType) => `${baseUrl}/${EmailType[type]}?chapterId=${chapterId}`,
  emails: (chapterId: string) => `${baseUrl}?chapterId=${chapterId}`,
  send: `${baseUrl}/send`
};

@Injectable({
  providedIn: 'root'
})
export class EmailAdminService {

  constructor(private http: HttpClient) { }

  addChapterEmailProvider(chapterId: string, provider: ChapterEmailProvider): Observable<ServiceResult<void>> {
    const params: HttpParams = HttpUtils.createFormParams({
      batchSize: provider.batchSize ? provider.batchSize.toString() : null,
      dailyLimit: provider.dailyLimit.toString(),
      fromEmailAddress: provider.fromEmailAddress,
      fromName: provider.fromName,
      smtpLogin: provider.smtpLogin,
      smtpPassword: provider.smtpPassword,
      smtpPort: provider.smtpPort.toString(),
      smtpServer: provider.smtpServer
    });

    return this.http.post(endpoints.chapterEmailProviders(chapterId), params).pipe(
      map((): ServiceResult<void> => ({
        success: true
      })),
      catchApiError()
    );
  }

  deleteChapterEmail(chapterId: string, type: EmailType): Observable<void> {
    return this.http.delete(endpoints.chapterEmail(chapterId, type)).pipe(
      map(() => undefined)
    );
  }

  deleteChapterEmailProvider(chapterId: string, chapterEmailProviderId: string): Observable<void> {
    return this.http.delete(endpoints.chapterEmailProvider(chapterId, chapterEmailProviderId)).pipe(
      map(() => undefined)
    );
  }

  getChapterAdminEmailProvider(chapterId: string, chapterEmailProviderId): Observable<ChapterEmailProvider> {
    return this.http.get(endpoints.chapterEmailProvider(chapterId, chapterEmailProviderId)).pipe(
      map((response: any) => this.mapChapterEmailProvider(response))
    );
  }

  getChapterAdminEmailProviders(chapterId: string): Observable<ChapterEmailProvider[]> {
    return this.http.get(endpoints.chapterEmailProviders(chapterId)).pipe(
      map((response: any) => response.map(x => this.mapChapterEmailProvider(x)))
    );
  }

  getChapterEmail(chapterId: string, type: EmailType): Observable<ChapterEmail> {
    return this.http.get(endpoints.chapterEmail(chapterId, type)).pipe(
      map((response: any) => this.mapChapterEmail(response))
    );
  }

  getChapterEmails(chapterId: string): Observable<ChapterEmail[]> {
    return this.http.get(endpoints.chapterEmails(chapterId)).pipe(
      map((response: any) => response.map(x => this.mapChapterEmail(x)))
    );
  }

  getEmail(currentChapterId: string, type: EmailType): Observable<Email> {
    return this.http.get(endpoints.email(currentChapterId, type)).pipe(
      map((response: any) => this.mapEmail(response))
    );
  }

  getEmails(currentChapterId: string): Observable<Email[]> {
    return this.http.get(endpoints.emails(currentChapterId)).pipe(
      map((response: any) => response.map(x => this.mapEmail(x)))
    );
  }

  sendEmail(memberId: string, subject: string, body: string): Observable<void> {
    const params: HttpParams = HttpUtils.createFormParams({
      body,
      memberId,
      subject
    });

    return this.http.post(endpoints.send, params).pipe(
      map(() => undefined)
    );
  }

  updateChapterAdminEmailProvider(chapterId: string, provider: ChapterEmailProvider): Observable<void> {
    const params: HttpParams = HttpUtils.createFormParams({
      batchSize: provider.batchSize ? provider.batchSize.toString() : null,
      dailyLimit: provider.dailyLimit.toString(),
      fromEmailAddress: provider.fromEmailAddress,
      fromName: provider.fromName,
      smtpLogin: provider.smtpLogin,
      smtpPassword: provider.smtpPassword,
      smtpPort: provider.smtpPort.toString(),
      smtpServer: provider.smtpServer
    });

    return this.http.put(endpoints.chapterEmailProvider(chapterId, provider.id), params).pipe(
      map(() => undefined)
    );
  }

  updateChapterEmail(chapterId: string, chapterEmail: ChapterEmail): Observable<void> {
    const params: HttpParams = HttpUtils.createFormParams({
      htmlContent: chapterEmail.htmlContent,
      subject: chapterEmail.subject
    });

    return this.http.put(endpoints.chapterEmail(chapterId, chapterEmail.type), params).pipe(
      map(() => undefined)
    );
  }

  updateEmail(email: Email, currentChapterId: string): Observable<void> {
    const params: HttpParams = HttpUtils.createFormParams({
      htmlContent: email.htmlContent,
      subject: email.subject
    });

    return this.http.put(endpoints.email(currentChapterId, email.type), params).pipe(
      map(() => undefined)
    );
  }

  private getEmailName(response: any): string {
    return StringUtils.camelPad(EmailType[response.type] || 'Unknown');
  }

  private mapChapterEmail(response: any): ChapterEmail {
    return {
      htmlContent: response.htmlContent,
      id: response.id,
      name: this.getEmailName(response),
      subject: response.subject,
      type: response.type
    };
  }

  private mapChapterEmailProvider(response: any): ChapterEmailProvider {
    return {
      batchSize: response.batchSize,
      dailyLimit: response.dailyLimit,
      fromEmailAddress: response.fromEmailAddress,
      fromName: response.fromName,
      id: response.id,
      order: response.order,
      smtpLogin: response.smtpLogin,
      smtpPassword: response.smtpPassword,
      smtpPort: response.smtpPort,
      smtpServer: response.smtpServer
    };
  }

  private mapEmail(response: any): Email {
    return {
      htmlContent: response.htmlContent,
      name: this.getEmailName(response),
      subject: response.subject,
      type: response.type
    };
  }
}

File: ODK.Website.Api/ODK.Services/Emails/EmailAdminService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using ODK.Core.Chapters;
using ODK.Core.Emails;
using ODK.Services.Exceptions;

namespace ODK.Services.Emails
{
    public class EmailAdminService : OdkAdminServiceBase, IEmailAdminService
    {
        private readonly IChapterRepository _chapterRepository;
        private readonly IEmailRepository _emailRepository;

        public EmailAdminService(IChapterRepository chapterRepository, IEmailRepository emailRepository)
            : base(chapterRepository)
        {
            _chapterRepository = chapterRepository;
            _emailRepository = emailRepository;
        }

        public async Task<ServiceResult> AddChapterEmailProvider(Guid currentMemberId, Guid chapterId, UpdateChapterEmailProvider provider)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            ChapterEmailProvider create = new ChapterEmailProvider(Guid.Empty, chapterId, provider.SmtpServer, provider.SmtpPort,
                provider.SmtpLogin, provider.SmtpPassword, provider.FromEmailAddress, provider.FromName,
                provider.BatchSize, provider.DailyLimit, 0);

            IReadOnlyCollection<ChapterEmailProvider> existing = await _chapterRepository.GetChapterEmailProviders(chapterId);

            create.Order = existing.Count + 1;

            ServiceResult validationResult = ValidateChapterEmailProvider(create);
            if (!validationResult.Success)
            {
                return validationResult;
            }

            await _chapterRepository.AddChapterEmailProvider(create);

            return ServiceResult.Successful();
        }

        public async Task DeleteChapterEmail(Guid currentMemberId, Guid chapterId, EmailType type)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            await _emailRepository.DeleteChapterEmail(chapterId, type);
        }

        public async Task<ServiceResult> DeleteChapterEmail(Guid currentMemberId, string chapterName, EmailType type)
        {
            Chapter chapter = await _chapterRepository.GetChapter(chapterName);
            if (chapter == null)
            {
                return ServiceResult.Failure("Chapter not found");
            }

            await AssertMemberIsChapterAdmin(currentMemberId, chapter.Id);

            await _emailRepository.DeleteChapterEmail(chapter.Id, type);

            return ServiceResult.Successful();
        }

        public async Task DeleteChapterEmailProvider(Guid currentMemberId, Guid chapterEmailProviderId)
        {
            ChapterEmailProvider provider = await GetChapterEmailProvider(currentMemberId, chapterEmailProviderId);

            await _chapterRepository.DeleteChapterEmailProvider(provider.Id);
        }

        public async Task<ChapterEmail> GetChapterEmail(Guid currentMemberId, Guid chapterId, EmailType type)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            ChapterEmail chapterEmail = await _emailRepository.GetChapterEmail(chapterId, type);
            if (chapterEmail != null)
            {
                return chapterEmail;
            }

            Email email = await _emailRepository.GetEmail(type, chapterId);
            return new ChapterEmail(Guid.Empty, chapterId, type, email.Subject, email.HtmlContent);
        }

        public async Task<ChapterEmailProvider> GetChapterEmailProvider(Guid currentMemberId, Guid chapterEmailProviderId)
        {
            ChapterEmailProvider provider = await _chapterRepository.GetChapterEmailProvider(chapterEmailProviderId);

            await AssertMemberIsChapterAdmin(currentMemberId, provider.ChapterId);

            return provider;
        }

        public async Task<IReadOnlyCollection<ChapterEmailProvider>> GetChapterEmailProviders(Guid currentMemberId, Guid chapterId)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            return await _chapterRepository.GetChapterEmailProviders(chapterId);
        }

        public async Task<IReadOnlyCollection<ChapterEmail>> GetChapterEmails(Guid currentMemberId, Guid chapterId)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            IReadOnlyCollection<ChapterEmail> chapterEmails = await _emailRepository.GetChapterEmails(chapterId);
            IDictionary<EmailType, ChapterEmail> chapterEmailDictionary = chapterEmails.ToDictionary(x => x.Type);

            List<ChapterEmail> defaultEmails = new List<ChapterEmail>();
            foreach (EmailType type in Enum.GetValues(typeof(EmailType)))
            {
                if (type == EmailType.None)
                {
                    continue;
                }

                if (!chapterEmailDictionary.ContainsKey(type))
                {
                    Email email = await _emailRepository.GetEmail(type, chapterId);
                    defaultEmails.Add(new ChapterEmail(Guid.Empty, chapterId, type, email.Subject, email.HtmlContent));
                }
            }

            return chapterEmails
                .Union(defaultEmails)
                .OrderBy(x => x.Type)
                .ToArray();
        }

        public async Task<Email> GetEmail(Guid currentMemberId, Guid currentChapterId, EmailType type)
        {
            await AssertMemberIsChapterSuperAdmin(currentMemberId, currentChapterId);

            return await _emailRepository.GetEmail(type);
        }

        public async Task<IReadOnlyCollection<Email>> GetEmails(Guid currentMemberId, Guid currentChapterId)
        {
            await AssertMemberIsChapterSuperAdmin(currentMemberId, currentChapterId);

            return await _emailRepository.GetEmails();
        }

        public async Task<ServiceResult> UpdateChapterEmail(Guid currentMemberId, Guid chapterId, EmailType type, UpdateEmail chapterEmail)
        {
            await AssertMemberIsChapterAdmin(currentMemberId, chapterId);

            ChapterEmail existing = await _emailRepository.GetChapterEmail(chapterId, type);
            if (existing == null)
            {
                existing = new ChapterEmail(Guid.Empty, chapterId, type, chapterEmail.Subject, chapterEmail.HtmlContent);
            }
            else
            {
                existing.HtmlContent = chapterEmail.HtmlContent;
                existing.Subject = chapterEmail.Subject;
            }

            ServiceResult validationResult = ValidateChapterEmail(existing);
            if (!validationResult.Success)
            {
                return validationResult;
            }

            if (existing.Id != Guid.Empty)
            {
                await _emailRepository.UpdateChapterEmail(existing);
            }
            else
            {
                await _emailRepository.AddChapterEmail(existing);
            }

            return ServiceResult.Successful();
        }

        public async Task<ServiceResult> UpdateChapterEmailProvider(Guid currentMemberId, Guid chapterEmailProviderId,
            UpdateChapterEmailProvider provider)
        {
            ChapterEmailProvider update = await GetChapterEmailProvider(currentMemberId, chapterEmailProviderId);

            update.BatchSize = provider.BatchSize;
            update.DailyLimit = provider.DailyLimit;
            update.FromEmailAddress = provider.FromEmailAddress;
            update.FromName = provider.FromName;
            update.SmtpLogin = provider.SmtpLogin;
            update.SmtpPassword = provider.SmtpPassword;
            update.SmtpPort = provider.SmtpPort;
            update.SmtpServer = provider.SmtpServer;

            ServiceResult validationResult = ValidateChapterEmailProvider(update);
            if (!validationResult.Success)
            {
                return validationResult;
            }

            await _chapterRepository.UpdateChapterEmailProvider(update);
            
            return ServiceResult.Successful();
        }

        public async Task UpdateEmail(Guid currentMemberId, Guid currentChapterId, EmailType type, UpdateEmail email)
        {
            await AssertMemberIsChapterSuperAdmin(currentMemberId, currentChapterId);

            Email existing = await _emailRepository.GetEmail(type);

            existing.HtmlContent = email.HtmlContent;
            existing.Subject = email.Subject;

            ValidateEmail(existing);

            await _emailRepository.UpdateEmail(existing);
        }

        private static void AssertChapterEmailValid(ChapterEmail chapterEmail)
        {
            ServiceResult result = ValidateChapterEmail(chapterEmail);
            if (!result.Success)
            {
                throw new OdkServiceException(result.Message);
            }
        }

        private static ServiceResult ValidateChapterEmail(ChapterEmail chapterEmail)
        {
            if (!Enum.IsDefined(typeof(EmailType), chapterEmail.Type) || chapterEmail.Type == EmailType.None)
            {
                return ServiceResult.Failure("Invalid type");
            }

            if (string.IsNullOrWhiteSpace(chapterEmail.HtmlContent) ||
                string.IsNullOrWhiteSpace(chapterEmail.Subject))
            {
                return ServiceResult.Failure("Some required fields are missing");
            }

            return ServiceResult.Successful();
        }

        private static void AssertChapterEmailProviderValid(ChapterEmailProvider provider)
        {
            ServiceResult result = ValidateChapterEmailProvider(provider);
            if (!result.Success)
            {
                throw new OdkServiceException("Some required fields are missing");
            }
        }

        private static ServiceResult ValidateChapterEmailProvider(ChapterEmailProvider provider)
        {
            if (string.IsNullOrWhiteSpace(provider.FromEmailAddress) ||
                string.IsNullOrWhiteSpace(provider.FromName) ||
                string.IsNullOrWhiteSpace(provider.SmtpLogin) ||
                string.IsNullOrWhiteSpace(provider.SmtpPassword) ||
                provider.SmtpPort == 0 ||
                provider.DailyLimit <= 0 ||
                provider.BatchSize <= 0)
            {
                return ServiceResult.Failure("Some required fields are missing");
            }

            return ServiceResult.Successful();
        }

        private static void ValidateEmail(Email email)
        {
            if (!Enum.IsDefined(typeof(EmailType), email.Type) || email.Type == EmailType.None)
            {
                throw new OdkServiceException("Invalid type");
            }

            if (string.IsNullOrWhiteSpace(email.HtmlContent) ||
                string.IsNullOrWhiteSpace(email.Subject))
            {
                throw new OdkServiceException("Some required fields are missing");
            }
        }
    }
}
